pkg/auth/rbac.go

```diff
@@ -0,0 +1,391 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package auth
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+	"kmesh.net/kmesh/api/v2/workloadapi"
+	"kmesh.net/kmesh/api/v2/workloadapi/security"
+)
+
+const (
+	SPIFFE_PREFIX = "spiffe://"
+)
+
+type Rbac struct {
+	policyStore *PolicyStore
+	workload    *workloadapi.Workload
+}
+
+func NewRbac() *Rbac {
+	return &Rbac{
+		policyStore: newPolicystore(),
+		workload:    &workloadapi.Workload{},
+	}
+}
+
+func (rbac *Rbac) DoRbac(conn *RbacConnection) bool {
+	allowPolices, denyPolicies := rbac.aggregate()
+
+	// 1. If there is ANY deny policy, deny the request
+	for _, denyPolicy := range denyPolicies {
+		if matches(conn, denyPolicy) {
+			return false
+		}
+	}
+
+	// 2. If there is NO allow policy for the workload, allow the request
+	if len(allowPolices) == 0 {
+		return true
+	}
+
+	// 3. If there is ANY allow policy matched, allow the request
+	for _, allowPolicy := range allowPolices {
+		if matches(conn, allowPolicy) {
+			return true
+		}
+	}
+
+	// 4. If 1,2 and 3 unsatisfied, deny the request
+	return false
+}
+
+func (rbac *Rbac) UpdatePolicy(auth *security.Authorization) error {
+	return rbac.policyStore.updatePolicy(auth)
+}
+
+func (rbac *Rbac) RemovePolicy(policyKey string) error {
+	return rbac.policyStore.removePolicy(policyKey)
+}
+
+func (rbac *Rbac) aggregate() (allowPolicies, denyPolicies []*AuthPolicy) {
+	// Collect policy names from workload, global namespace and namespace
+	policyNames := make([]string, len(rbac.workload.GetAuthorizationPolicies()))
+	copy(policyNames, rbac.workload.GetAuthorizationPolicies())
+	policyNames = append(policyNames, strSetToSlice(rbac.policyStore.byNamespace[""])...)
+	policyNames = append(policyNames, strSetToSlice(rbac.policyStore.byNamespace[rbac.workload.Namespace])...)
+
+	allowPolicies = make([]*AuthPolicy, 0)
+	denyPolicies = make([]*AuthPolicy, 0)
+	for _, name := range policyNames {
+		if policy, ok := rbac.policyStore.byKey[name]; !ok {
+			continue
+		} else {
+			if policy.Action == security.Action_ALLOW {
+				allowPolicies = append(allowPolicies, policy)
+			} else if policy.Action == security.Action_DENY {
+				denyPolicies = append(denyPolicies, policy)
+			}
+		}
+	}
+	return
+}
+
+func matches(conn *RbacConnection, policy *AuthPolicy) bool {
+	if policy.GetRules() == nil {
+		return false
+	}
+
+	eg := &errgroup.Group{}
+	mut := &sync.Mutex{}
+	// If ANY rule matches, it's a match
+	for _, rule := range policy.GetRules() {
+		ruleMatch := true
+		// If ALL clause matches, it's a match
+		for _, clause := range rule.GetClauses() {
+			clauseMatch := false
+			// If ANY match matches, it's a match
+			for _, match := range clause.GetMatches() {
+				if isEmptyMatch(match) {
+					continue
+				}
+
+				// Values of specific type are OR-ed. If multiple types are set, they are AND-ed
+				anyMatch := true
+				eg.Go(func() error {
```

Comment:
These 5 goroutines can be early exited if anyMatch turns out to false in any goroutine

Reply:
Ztunnel uses a serial match here, since we are using a parallel match, and the performance should be better than Ztunnel, therefore we may not need an early exit for now